### PR TITLE
Made OpticalDuplicateFinder Serializable

### PR DESCRIPTION
**Pull request by jamesemery**
In tie-out for MarkDuplicatesSpark we found there were issues passing the proper read name regex arguments to the `OpticalDuplicateFinder` because the owned fields are private and the classes themselves aren't serializable resulting the fields being reset to their default value whenever they are passed around. In order to work around this it is easiest and most future proof to make the corresponding classes in Picard (which MarkDuplicatesSpark calls into) `Serializable`so the default java field serializer will take care of the fields for us.

### CODE REVIEW

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 1]
```diff
@@ -190,4 +194,19 @@ public static int rapidParseInt(final String input) throws NumberFormatException
...
     }
+
+
+    private void readObject(ObjectInputStream aInputStream) throws ClassNotFoundException, IOException {
+        // perform the default de-serialization first
+        aInputStream.defaultReadObject();
+
+        // Ensure that if the regex is the default
+        if (this.readNamePattern.equals(DEFAULT_READ_NAME_REGEX)) {
+            readNamePattern.equals(DEFAULT_READ_NAME_REGEX);
```
Comment (yfarjoun) [inline src/main/java/picard/sam/util/ReadNameParser.java]: what is the purpose of this statement?

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 2]
```diff
@@ -26,15 +30,15 @@
...
-    private final int[] tmpLocationFields = new int[3]; // for optimization of addLocationInformation
+    private transient final int[] tmpLocationFields = new int[3]; // for optimization of addLocationInformation
 
     private String readNameRegex = null;
 
-    private Pattern readNamePattern;
+    private transient Pattern readNamePattern;
 
-    private boolean warnedAboutRegexNotMatching = false;
+    private transient boolean warnedAboutRegexNotMatching = false;
```
Comment (lbergelson) [inline src/main/java/picard/sam/util/ReadNameParser.java]: This doesn't seem like it should be transient.

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 3]
```diff
@@ -26,15 +30,15 @@
...
 
-    private final Log log;
+    private transient final Log log;
```
Comment (lbergelson) [inline src/main/java/picard/sam/util/ReadNameParser.java]: Make sure this is reinitialized by the deserializer if necessary.

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 4]
```diff
@@ -26,15 +30,15 @@
      */
     public static final String DEFAULT_READ_NAME_REGEX = "<optimized capture of last three ':' separated fields as numeric values>".intern();
 
-    private final int[] tmpLocationFields = new int[3]; // for optimization of addLocationInformation
+    private transient final int[] tmpLocationFields = new int[3]; // for optimization of addLocationInformation
```
Comment (lbergelson) [inline src/main/java/picard/sam/util/ReadNameParser.java]: I know it's probably unecessary to serialize this because it's a tmp field, but this is not a commonly serialized thing so shaving a few bytes off isn't worth the complexity of unnecessary transitivity.   I would make this non-transient, or if it must be transient I believe it has to be reinitialized in the deserializer.

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 5]
```diff
@@ -26,15 +30,15 @@
...
 
     private String readNameRegex = null;
 
-    private Pattern readNamePattern;
+    private transient Pattern readNamePattern;
```
Comment (lbergelson) [inline src/main/java/picard/sam/util/ReadNameParser.java]: Is this not serializable or just an optimization to reduce serialization size?  If it's the second case I'd make it non-transient.

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 6]
```diff
@@ -190,4 +194,19 @@ public static int rapidParseInt(final String input) throws NumberFormatException
         if (isNegative) val = -val;
         return val;
     }
+
+
+    private void readObject(ObjectInputStream aInputStream) throws ClassNotFoundException, IOException {
+        // perform the default de-serialization first
+        aInputStream.defaultReadObject();
+
+        // Ensure that if the regex is the default
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/picard/sam/util/ReadNameParser.java]: Maybe rephrase as "Object equality is used when comparing the regex strings, but is not maintained through serialization, check if the serialized regex is `equal` to the default regex and then replace it with the canonical string if it is."

File: src/main/java/picard/sam/util/ReadNameParser.java (hunk 1 above)
Comment (lbergelson) [reply src/main/java/picard/sam/util/ReadNameParser.java]: `equals` -> `=`

**Comment by jamesemery** [comment]
@lbergelson @yfarjoun I have responded to your comments and added a test of the readname parser serialization behavior.

File: src/main/java/picard/sam/util/ReadNameParser.java [hunk 7]
```diff
@@ -64,6 +63,7 @@ public ReadNameParser(final String readNameRegex) {
      * @param log the log to which to write messages.
      */
     public ReadNameParser(final String readNameRegex, final Log log) {
+        this.useOptimizedDefaultParsing = readNameRegex.equals(DEFAULT_READ_NAME_REGEX);
```
Comment (yfarjoun) [inline src/main/java/picard/sam/util/ReadNameParser.java]: still needs to be ==

File: src/test/java/picard/sam/util/ReadNameParserTests.java [hunk 8]
```diff
@@ -139,4 +140,46 @@ else if (readNameRegex == ReadNameParser.DEFAULT_READ_NAME_REGEX) { // additiona
...
+    public void testSerializedReadNameParser () throws IOException, ClassNotFoundException {
+        // A non standard parser
+        final String lastThreeFieldsRegex = "(?:.*:)?([0-9]+)[^:]*:([0-9]+)[^:]*:([0-9]+)[^:]*$";
+        final ReadNameParser parser = new ReadNameParser(lastThreeFieldsRegex);
+        final ReadNameParser defaultParser = new ReadNameParser();
+        PhysicalLocationInt loc = new PhysicalLocationInt();
+        String readName = "1109ABC:22981DEF:17995GHI";
+
+        Assert.assertTrue(parser.addLocationInformation(readName, loc));
+        Assert.assertEquals(loc.getTile(), 1109);
```
Comment (yfarjoun) [inline src/test/java/picard/sam/util/ReadNameParserTests.java]: make constants from numbers.

File: src/test/java/picard/sam/util/ReadNameParserTests.java [hunk 9]
```diff
@@ -139,4 +140,46 @@ else if (readNameRegex == ReadNameParser.DEFAULT_READ_NAME_REGEX) { // additiona
...
+            loc = new PhysicalLocationInt();
+            Assert.assertTrue(parserSerialized.addLocationInformation(readName, loc));
+            Assert.assertEquals(loc.getTile(), 1109);
+            Assert.assertEquals(loc.getX(), 22981);
+            Assert.assertEquals(loc.getY(), 17995);
+
+            loc = new PhysicalLocationInt();
+            Assert.assertFalse(defaultParserSerialized.addLocationInformation(readName, loc));
+        }
+
```
Comment (yfarjoun) [inline src/test/java/picard/sam/util/ReadNameParserTests.java]: remove extra newlines

File: src/test/java/picard/sam/util/ReadNameParserTests.java [hunk 10]
```diff
@@ -139,4 +140,46 @@ else if (readNameRegex == ReadNameParser.DEFAULT_READ_NAME_REGEX) { // additiona
...
+        final byte[] serializedBytes;
+        try(final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            final ObjectOutputStream out = new ObjectOutputStream(baos)){
+            out.writeObject(parser);
+            out.writeObject(defaultParser);
+            serializedBytes = baos.toByteArray();
+        }
+        try (final ByteArrayInputStream bais = new ByteArrayInputStream(serializedBytes);
+             final ObjectInputStream in = new ObjectInputStream(bais)){
```
Comment (yfarjoun) [inline src/test/java/picard/sam/util/ReadNameParserTests.java]: fix indentations and space

File: src/test/java/picard/sam/util/ReadNameParserTests.java [hunk 11]
```diff
@@ -139,4 +140,46 @@ else if (readNameRegex == ReadNameParser.DEFAULT_READ_NAME_REGEX) { // additiona
...
 
+    @Test
+    // Testing that the parser behavior stays constant after being java serialized
+    public void testSerializedReadNameParser () throws IOException, ClassNotFoundException {
+        // A non standard parser
+        final String lastThreeFieldsRegex = "(?:.*:)?([0-9]+)[^:]*:([0-9]+)[^:]*:([0-9]+)[^:]*$";
+        final ReadNameParser parser = new ReadNameParser(lastThreeFieldsRegex);
+        final ReadNameParser defaultParser = new ReadNameParser();
+        PhysicalLocationInt loc = new PhysicalLocationInt();
+        String readName = "1109ABC:22981DEF:17995GHI";
```
Comment (yfarjoun) [inline src/test/java/picard/sam/util/ReadNameParserTests.java]: final readname

File: src/test/java/picard/sam/util/ReadNameParserTests.java [hunk 12]
```diff
@@ -139,4 +140,46 @@ else if (readNameRegex == ReadNameParser.DEFAULT_READ_NAME_REGEX) { // additiona
...
+        final byte[] serializedBytes;
+        try(final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            final ObjectOutputStream out = new ObjectOutputStream(baos)){
+            out.writeObject(parser);
+            out.writeObject(defaultParser);
+            serializedBytes = baos.toByteArray();
+        }
+        try (final ByteArrayInputStream bais = new ByteArrayInputStream(serializedBytes);
+             final ObjectInputStream in = new ObjectInputStream(bais)){
+            ReadNameParser parserSerialized = (ReadNameParser) in.readObject();
```
Comment (yfarjoun) [inline src/test/java/picard/sam/util/ReadNameParserTests.java]: can be final x2

**Comment by jamesemery** [comment]
@yfarjoun I have fixed the issues you pointed out. Do you have any other comments for this branch?

File: src/main/java/picard/sam/util/ReadNameParser.java (hunk 7 above)
Comment (yfarjoun) [reply src/main/java/picard/sam/util/ReadNameParser.java]: right. got confused.

**Comment by yfarjoun** [comment]
👍
